File: common_libs/src/main/java/io/virgo_common/common_libs/customView/shimmer/Shimmer.java
package io.virgo_common.common_libs.customView.shimmer;

import android.animation.Animator;
import android.animation.ObjectAnimator;
import android.animation.ValueAnimator;
import android.view.View;

import androidx.annotation.NonNull;

/**
 * Shimmer
 * User: romainpiel
 * Date: 06/03/2014
 * Time: 15:42
 */
public class Shimmer {

    public static final int ANIMATION_DIRECTION_LTR = 0;
    public static final int ANIMATION_DIRECTION_RTL = 1;

    private static final int DEFAULT_REPEAT_COUNT = ValueAnimator.INFINITE;
    private static final long DEFAULT_DURATION = 1000;
    private static final long DEFAULT_START_DELAY = 0;
    private static final int DEFAULT_DIRECTION = ANIMATION_DIRECTION_LTR;

    private int repeatCount;
    private long duration;
    private long startDelay;
    private int direction;
    private Animator.AnimatorListener animatorListener;

    private ObjectAnimator animator;

    public Shimmer() {
        repeatCount = DEFAULT_REPEAT_COUNT;
        duration = DEFAULT_DURATION;
        startDelay = DEFAULT_START_DELAY;
        direction = DEFAULT_DIRECTION;
    }

    public int getRepeatCount() {
        return repeatCount;
    }

    public Shimmer setRepeatCount(int repeatCount) {
        this.repeatCount = repeatCount;
        return this;
    }

    public long getDuration() {
        return duration;
    }

    public Shimmer setDuration(long duration) {
        this.duration = duration;
        return this;
    }

    public long getStartDelay() {
        return startDelay;
    }

    public Shimmer setStartDelay(long startDelay) {
        this.startDelay = startDelay;
        return this;
    }

    public int getDirection() {
        return direction;
    }

    public Shimmer setDirection(int direction) {

        if (direction != ANIMATION_DIRECTION_LTR && direction != ANIMATION_DIRECTION_RTL) {
            throw new IllegalArgumentException("The animation direction must be either ANIMATION_DIRECTION_LTR or ANIMATION_DIRECTION_RTL");
        }

        this.direction = direction;
        return this;
    }

    public Animator.AnimatorListener getAnimatorListener() {
        return animatorListener;
    }

    public Shimmer setAnimatorListener(Animator.AnimatorListener animatorListener) {
        this.animatorListener = animatorListener;
        return this;
    }

    public <V extends View & ShimmerViewBase> void start(final V shimmerView) {

        if (isAnimating()) {
            return;
        }

        final Runnable animate = () -> {

            shimmerView.setShimmering(true);

            float fromX = 0;
            float toX = shimmerView.getWidth();
            if (direction == ANIMATION_DIRECTION_RTL) {
                fromX = shimmerView.getWidth();
                toX = 0;
            }

            animator = ObjectAnimator.ofFloat(shimmerView, "gradientX", fromX, toX);
            animator.setRepeatCount(repeatCount);
            animator.setDuration(duration);
            animator.setStartDelay(startDelay);
            animator.addListener(new Animator.AnimatorListener() {
                @Override
                public void onAnimationStart(@NonNull Animator animation) {
                }

                @Override
                public void onAnimationEnd(@NonNull Animator animation) {
                    shimmerView.setShimmering(false);

                    shimmerView.postInvalidateOnAnimation();

                    animator = null;
                }

                @Override
                public void onAnimationCancel(@NonNull Animator animation) {

                }

                @Override
                public void onAnimationRepeat(@NonNull Animator animation) {

                }
            });

            if (animatorListener != null) {
                animator.addListener(animatorListener);
            }

            animator.start();
        };

        if (!shimmerView.isSetUp()) {
            shimmerView.setAnimationSetupCallback(target -> animate.run());
        } else {
            animate.run();
        }
    }

    public void cancel() {
        if (animator != null) {
            animator.cancel();
        }
    }

    public boolean isAnimating() {
        return animator != null && animator.isRunning();
    }
}

File: common_libs/src/main/java/io/virgo_common/common_libs/customView/shimmerK/Shimmer.kt
package io.virgo_common.common_libs.customView.shimmerK

import android.animation.Animator
import android.animation.ObjectAnimator
import android.animation.ValueAnimator
import android.view.View
import io.virgo_common.common_libs.customView.shimmerK.ShimmerViewHelper.AnimationSetupCallback

/**
 * Shimmer
 * User: romainpiel
 * Date: 06/03/2014
 * Time: 15:42
 *
 * Source: `https://github.com/romainpiel/Shimmer-android`
 */
class Shimmer {
	var repeatCount: Int
		private set
	var duration: Long
		private set
	var startDelay: Long
		private set
	var direction: Int
		private set
	var animatorListener: Animator.AnimatorListener? = null
		private set

	private var animator: ObjectAnimator? = null

	init {
		repeatCount = DEFAULT_REPEAT_COUNT
		duration = DEFAULT_DURATION
		startDelay = DEFAULT_START_DELAY
		direction = DEFAULT_DIRECTION
	}

	fun setRepeatCount(repeatCount: Int): Shimmer {
		this.repeatCount = repeatCount
		return this
	}

	fun setDuration(duration: Long): Shimmer {
		this.duration = duration
		return this
	}

	fun setStartDelay(startDelay: Long): Shimmer {
		this.startDelay = startDelay
		return this
	}

	fun setDirection(direction: Int): Shimmer {
		require(!(direction != ANIMATION_DIRECTION_LTR && direction != ANIMATION_DIRECTION_RTL)) { "The animation direction must be either ANIMATION_DIRECTION_LTR or ANIMATION_DIRECTION_RTL" }

		this.direction = direction
		return this
	}

	fun setAnimatorListener(animatorListener: Animator.AnimatorListener?): Shimmer {
		this.animatorListener = animatorListener
		return this
	}

	fun <V> start(shimmerView: V) where V : View, V : ShimmerViewBase? {
		if (this.isAnimating) {
			return
		}

		val animate: Runnable = object : Runnable {
			override fun run() {
				shimmerView.isShimmering = true

				var fromX = 0f
				var toX = shimmerView.width.toFloat()
				if (direction == ANIMATION_DIRECTION_RTL) {
					fromX = shimmerView.width.toFloat()
					toX = 0f
				}

				animator = ObjectAnimator.ofFloat(shimmerView, "gradientX", fromX, toX)
				animator!!.repeatCount = repeatCount
				animator!!.duration = duration
				animator!!.startDelay = startDelay
				animator!!.addListener(object : Animator.AnimatorListener {
					override fun onAnimationStart(
						animation: Animator,
						isReverse: Boolean
					) {
						super.onAnimationStart(animation, isReverse)
					}

					override fun onAnimationStart(p0: Animator) {

					}

					override fun onAnimationEnd(
						animation: Animator,
						isReverse: Boolean
					) {
						super.onAnimationEnd(animation, isReverse)
					}

					override fun onAnimationEnd(p0: Animator) {
						shimmerView.isShimmering = false

						shimmerView.postInvalidateOnAnimation()

						animator = null
					}

					override fun onAnimationCancel(p0: Animator) {

					}

					override fun onAnimationRepeat(p0: Animator) {

					}
				})

				if (animatorListener != null) {
					animator!!.addListener(animatorListener)
				}

				animator!!.start()
			}
		}

		if (!shimmerView.isSetUp) {
			shimmerView.setAnimationSetupCallback(object : AnimationSetupCallback {
				override fun onSetupAnimation(target: View?) {
					animate.run()
				}
			})
		} else {
			animate.run()
		}
	}

	fun cancel() {
		if (animator != null) {
			animator!!.cancel()
		}
	}

	val isAnimating: Boolean
		get() = animator != null && animator!!.isRunning

	companion object {
		const val ANIMATION_DIRECTION_LTR: Int = 0
		const val ANIMATION_DIRECTION_RTL: Int = 1

		const val DEFAULT_REPEAT_COUNT = ValueAnimator.INFINITE
		const val DEFAULT_DURATION: Long = 1000
		const val DEFAULT_START_DELAY: Long = 0
		const val DEFAULT_DIRECTION: Int = ANIMATION_DIRECTION_LTR
	}
}
